### PR TITLE
Separate different uses of minimum fees

### PR DESCRIPTION
**Pull request by morcos**
I've been meaning to do this for a while.
There is no behavior change here, but there were unintended consequences of having these all of these different notions linked to minRelayTxFee.

- blockmintxfee should be set by miners to reflect their marginal cost of transmitting extra bytes.
- incrementalrelayfee was previously hard-coded to the default minRelayTxFee, but it should be configurable so it can be adjusted to reflect the cost of network wide relay of transactions.
- dustrelayfee should not be accidentally changed when someone wants to limit their mempool to higher fee transactions.

I plan a future change to make the wallet smarter about not generating change outputs near the dust limit at all.  I also think we should consider splitting dust into 2 dust limits if we ever raise it again and first raise the creation limit and second the standard limit a version later.  Since it is not clear we'll ever need to increase the dust limit, this change isn't introduced now.

### CODE REVIEW

**Comment by gmaxwell** [comment]
utACK.  My comment on reading the code was that we need to split the wallet and non-wallet behavior (esp for dust) but I see that your PR comment mentions an intention to do that later.

**Comment by morcos** [comment]
rebased

**Comment by jtimon** [comment]
utACK 947b63a
If it is decided to hide some or all of the new options and only expose them after 0.14 branches for translation reasons that's fine with me too (although I don't see how no documentation is better than documentation in English).

**Comment by instagibbs** [comment]
utACK 947b63a06043480a3a9f490b0d5485486f73e5bd

File: src/init.cpp [hunk 1]
```diff
@@ -465,8 +465,10 @@ std::string HelpMessage(HelpMessageMode mode)
     AppendParamsHelpMessages(strUsage, showDebug);
 
     strUsage += HelpMessageGroup(_("Node relay options:"));
-    if (showDebug)
+    if (showDebug) {
         strUsage += HelpMessageOpt("-acceptnonstdtxn", strprintf("Relay and mine \"non-standard\" transactions (%sdefault: %u)", "testnet/regtest only; ", !Params(CBaseChainParams::TESTNET).RequireStandard()));
+        strUsage += HelpMessageOpt("-incrementalrelayfee=<amt>", strprintf(_("Fee rate (in %s/kB) used to define cost of free relay, used for mempool limiting and BIP 125 replacement. (default: %s)"), CURRENCY_UNIT, FormatMoney(DEFAULT_INCREMENTAL_RELAY_FEE)));
```
Comment (sdaftuar) [inline src/init.cpp]: Is this not supposed to be translated?  Also, "cost of free relay" is a very confusing phrase.  Perhaps "cost of relay"?

File: src/txmempool.h [hunk 2]
```diff
@@ -591,7 +586,7 @@ class CTxMemPool
 
     /** The minimum fee to get into the mempool, which may itself not be enough
       *  for larger-sized transactions.
-      *  The minReasonableRelayFee constructor arg is used to bound the time it
+      *  The incrementalRelayFee policy variable is used to bound the time it
```
Comment (sdaftuar) [inline src/txmempool.h]: Should still mention what minReasonableRelayFee constructor arg is for --  I guess fee estimation wants the min relay fee?

File: src/init.cpp [hunk 3]
```diff
@@ -468,6 +468,7 @@ std::string HelpMessage(HelpMessageMode mode)
     if (showDebug) {
         strUsage += HelpMessageOpt("-acceptnonstdtxn", strprintf("Relay and mine \"non-standard\" transactions (%sdefault: %u)", "testnet/regtest only; ", !Params(CBaseChainParams::TESTNET).RequireStandard()));
         strUsage += HelpMessageOpt("-incrementalrelayfee=<amt>", strprintf(_("Fee rate (in %s/kB) used to define cost of free relay, used for mempool limiting and BIP 125 replacement. (default: %s)"), CURRENCY_UNIT, FormatMoney(DEFAULT_INCREMENTAL_RELAY_FEE)));
+        strUsage += HelpMessageOpt("-dustrelayfee=<amt>", strprintf(_("Fee rate (in %s/kB) used to defined dust, the size of an output such that it will cost about 1/3 of its value in fees at this fee rate to spend it. (default: %s)"), CURRENCY_UNIT, FormatMoney(DUST_RELAY_TX_FEE)));
```
Comment (sdaftuar) [inline src/init.cpp]: This should not be translated?  Also, "size of an output" -> "value of an output"?

**Review by sdaftuar — CHANGES_REQUESTED** [review]
Couple nits, concept ACK, haven't tested yet (but plan to).

**Comment by sipa** [comment]
Concept ACK

File: src/txmempool.h (hunk 2 above)
Comment (morcos) [inline src/txmempool.h]: Removed in #9548 actually, but I will address your other points...

File: src/init.cpp (hunk 3 above)
Comment (MarcoFalke) [inline src/init.cpp]: This should not be called dust*relay*fee, but dustfee. My reasoning is that dustfee has more implications than just setting a relay fee rate.

**Comment by MarcoFalke** [comment]
ACK, but needs all debug strings removed from the translations.

**Comment by morcos** [comment]
Removed translation strings for -debug options.

Ideally this would get merged in time to keep the translation of `-blockmintxfee`.  I think it is ready for merge other than any disagreement over the name of `-dustrelayfee`.  I'm happy to change it if we agree on something else.     `-dustminfee`, `-dustdefinefee`, `-dustfee` or just leave it?

**Comment by laanwj** [comment]
utACK eb30d1a

**Pull request merged**